README.md
# node-bitx
[![Build Status](https://travis-ci.org/bausmeier/node-bitx.png)](https://travis-ci.org/bausmeier/node-bitx)
[![codecov](https://codecov.io/gh/bausmeier/node-bitx/branch/master/graph/badge.svg)](https://codecov.io/gh/bausmeier/node-bitx)

A simple wrapper for the Luno API. The module supports promise and callbacks.

## Usage
Add bitx as a dependency:

```bash
$ npm install --save bitx
```

### BitX([keyId, keySecret][options])
To access the private Luno API methods you must supply your key id and key secret as the first two arguments. If you are only accessing the public API endpoints you can leave these two arguments out.

The optional `options` argument can be used to override the default options. The default options are equivalent to:

```javascript
{
  hostname: 'api.luno.com',
  port: 443,
  ca: undefined,
  pair: 'XBTZAR'
}
```

## Properties

### apiCallRate

Gives the current rate per minute of API calls. The property is read only.
It can be used by applications to limit the API call rate to prevent HTTP error code `429` (too many requests) responses from Luno server.
See Luno API documentation for applicable rate limitations: https://www.luno.com/en/developers/api#tag/Rate-Limiting

> API calls which result in error code `429` or `ErrTooManyRequests` are considered failed/unprocessed server request and are not added to the `apiCallRate` counter

Example:

```javascript
console.log(bitx.apiCallRate)
```


## Methods
For details about the API endpoints see https://www.luno.com/en/developers/api.

### Callbacks
The arguments passed to the callback function for each method are:

1. An error or `null` if no error occurred.
1. An object containing the data returned by the Luno API.

The `callback` function is optional. If the `callback` is not provided, the methods return a promise. 

### getTicker([options][, callback])
GET https://api.luno.com/api/1/ticker/XBTZAR

Default options:

```javascript
{
  pair: bitx.pair
}
```

Example:

```javascript
bitx.getTicker(function(err, ticker) {});
```

### getAllTickers([callback])
GET https://api.luno.com/api/1/tickers

Example:

```javascript
bitx.getAllTickers(function(err, tickers) {});
```

### getOrderBook([options][, callback])
GET https://api.luno.com/api/1/orderbook

Default options:

```javascript
{
  pair: bitx.pair
}
```

Example:

```javascript
bitx.getOrderBook(function(err, orderBook) {});
```

### getTrades([options][, callback])
GET https://api.luno.com/api/1/trades

Default options:

```javascript
{
  pair: bitx.pair
}
```

Example:

```javascript
bitx.getTrades(function(err, trades) {});
```

### getTradeList([options][, callback])
GET https://api.luno.com/api/1/listtrades

Default options:

```javascript
{
  pair: bitx.pair
}
```

Example:

```javascript
bitx.getTradeList({sort_desc: true, limit: 10}, function(err, tradeList) {});
```

### getOrderList([options][, callback])
GET https://api.luno.com/api/1/listorders

Default options:

```javascript
{
  pair: bitx.pair,
  state: undefined
}
```

Example:

```javascript
bitx.getOrderList({state: 'PENDING'}, function(err, orderList) {});
```

### getOrderListV2([options][, callback])
GET https://api.luno.com/api/exchange/2/listorders

Default options:

```javascript
{
  pair: bitx.pair,
  state: undefined
}
```

Example:

```javascript
bitx.getOrderListV2({ closed: true }, function(err, orderList) {});
```

### getOrderListV3(options[, callback])
GET https://api.luno.com/api/exchange/3/order

Example:

```javascript
bitx.getOrderListV3({id: 'BXMC2CJ7HNB88U4' }, function(err, orderList) {});
```

### getBalance([asset][, callback])
GET https://api.luno.com/api/1/balance

Example:

```javascript
bitx.getBalance('ZAR', function(err, balance) {});
```

### getFundingAddress(asset[, options][, callback])
GET https://api.luno.com/api/1/funding_address

Default options:

```javascript
{
  address: undefined
}
```

Example:

```javascript
bitx.getFundingAddress('XBT', {address: 'B1tC0InExAMPL3fundIN6AdDreS5t0Use'}, function(err, fundingAddress) {});
```

### createFundingAddress(asset[, callback])
POST https://api.luno.com/api/1/funding_address

Example:

```javascript
bitx.createFundingAddress('XBT', function(err, fundingAddress) {});
```

### getFeeInfo([options][, callback])
GET https://api.luno.com/api/1/fee_info

Default options:

```javascript
{
  pair: bitx.pair
}
```

Example:

```javascript
bitx.getFeeInfo({pair: 'XBTZAR'}, function(err, feeInfo) {});
```

### postBuyOrder(volume, price[, options][, callback])
POST https://api.luno.com/api/1/postorder

Example:

```javascript
bitx.postBuyOrder(9999.99, 0.01, function(err, order) {});
```

### postSellOrder(volume, price[, options][, callback])
POST https://api.luno.com/api/1/postorder

Example:

```javascript
bitx.postSellOrder(0.01, 9999.99, function(err, order) {});
```

### postMarketBuyOrder(volume[, options][, callback])
POST https://api.luno.com/api/1/marketorder

Example:

```javascript
bitx.postMarketBuyOrder(0.01, function(err, order) {});
```

### postMarketSellOrder(volume[, options][, callback])
POST https://api.luno.com/api/1/marketorder

Example:

```javascript
bitx.postMarketSellOrder(0.01, function(err, order) {});
```

### stopOrder(orderId[, callback])
POST https://api.luno.com/api/1/stoporder

Example:

```javascript
bitx.stopOrder('BXMC2CJ7HNB88U4', function(err, result) {});
```

### getOrder(orderId[, callback])
GET https://api.luno.com/api/1/orders/{orderId}

Example:

```javascript
bitx.getOrder('BXHW6PFRRXKFSB4', function(err, result) {});
```

### getOrderV2(orderId[, callback])
GET https://api.luno.com/api/exchange/2/orders/{orderId}

Example:

```javascript
bitx.getOrder('BXHW6PFRRXKFSB4', function(err, result) {});
```

### getTransactions(asset[, options][, callback])
GET https://api.luno.com/api/1/transactions

Default options:
```javascript
{
  offset: 0,
  limit: 10
}
```

Example:

```javascript
bitx.getTransactions('XBT', {offset: 5, limit: 20}, function(err, transactions) {});
```

### getWithdrawals([callback])
GET https://api.luno.com/api/1/withdrawals

Example:

```javascript
bitx.getWithdrawals(function(err, withdrawals) {});
```

### getWithdrawal(withdrawalId[, callback])
GET https://api.luno.com/api/1/withdrawals/{withdrawalId}

Example:

```javascript
bitx.getWithdrawal('1212', function(err, withdrawal) {});
```

### requestWithdrawal(type, amount[, callback])
POST https://api.luno.com/api/1/withdrawals

Example:

```javascript
bitx.requestWithdrawal('ZAR_EFT', 1000, function(err, withdrawal) {});
```

### cancelWithdrawal(withdrawalId[, callback])
DELETE https://api.luno.com/api/1/withdrawals/{withdrawalId}

Example:

```javascript
bitx.cancelWithdrawal('1212', function(err, withdrawal) {});
```

## Contributing

Like my work? Please donate **1E1sebnWax5Br2mp8y9dox6oX9Snmf42uz**.

Don't like it? Open a pull request or create an issue and help me improve it.


CHANGELOG.md
### 1.10.0 (2022-06-20)

* Add `apiCallRate` property
* Update tests

### 1.9.1 (2022-06-20)

* Fix https request error handling when no `callback` parameter (promises)

### 1.9.0 (2021-12-06)

* Add methods getOrderListV2, getOrderV2, getOrderV3

### 1.8.0 (2020-10-13)

* Add promise support for all methods. When `callback` parameter is not provided, the methods return a promise

### 1.7.1

* Parse Luno API errors
* Update domain to `api.luno.com`

### 1.7.0

* Add getTradeList method

### 1.6.0

* Add options to postBuyOrder, postMarketBuyOrder, postSellOrder and postMarketSellOrder
* Add getFeeInfo method

### 1.5.1

* Fix an issue with GET requests sent after POST requests

### 1.5.0

* Add getOrder method

### 1.4.2

* Make asset parameter optional for getBalance

### 1.4.1

* Only catch JSON.parse errors (#6)

### 1.4.0

* Add pair option to getTicker, getOrderBook and getTrades
* Add getAllTickers method

### 1.3.0

* Add createFundingAddress method
* Add getTransactions method
* Accept address option for getFundingAddress
* Accept state option for getOrderList
* Add getWithdrawals method
* Add getWithdrawal method
* Add requestWithdrawal method
* Add cancelWithdrawal method

### 1.2.1

* Fix path bug (#1)

### 1.2.0

* Add getFundingAddress method

### 1.1.0

* Support multiple currency pairs
* BitX.getLimits is now deprecated (use BitX.getBalance instead).
* Updated to the latest BitX API.
* Added BitX.getBalance method.

### 1.0.1

* Set response encoding to utf8
* Resolve require path properly


test/BitXTests.js
'use strict';

var BitX = require('../lib/BitX'),
    https = require('https'),
    sinon = require('sinon'),
    chai = require('chai'),
    FakeRequest = require('./helpers/FakeRequest');

var expect = chai.expect;

describe('Constructor', function() {
  
  it('should create a new instance', function() {
    var bitx = new BitX();
    expect(bitx).to.be.an.instanceOf(BitX);
  });
  
  it('should create a new instance without the new keyword', function() {
    var bitx = BitX();
    expect(bitx).to.be.an.instanceOf(BitX);
  });
  
  it('should have default options', function() {
    var bitx = new BitX();
    expect(bitx.hostname).to.equal('bitx.co.za');
    expect(bitx.port).to.equal(443);
  });
  
  it('should accept options', function() {
    var options = {
      hostname: 'localhost',
      port: 8000
    };
    var bitx = new BitX(options);
    expect(bitx.hostname).to.equal(options.hostname);
    expect(bitx.port).to.equal(options.port);
  });
  
  it('should accept auth and options', function() {
    var keyId = 'cnz2yjswbv3jd',
        keySecret = '0hydMZDb9HRR3Qq-iqALwZtXLkbLR4fWxtDZvkB9h4I',
        options = {
          hostname: 'localhost',
          port: 8000,
        };
    var bitx = new BitX(keyId, keySecret, options);
    expect(bitx.hostname).to.equal(options.hostname);
    expect(bitx.port).to.equal(options.port);
    expect(bitx.auth).to.equal(keyId + ':' + keySecret);
  });
});

describe('Internal', function() {
  
  var bitx,
      request,
      stub;

  var keyId = '12345',
      keySecret = '0000000000000000',
      path = 'test';
  
  var expectedOptions = {
    headers: {
      'Accept': 'application/json',
      'Accept-Charset': 'utf-8'
    },
    hostname: 'bitx.co.za',
    path: '/api/1/BTCZAR/' + path,
    port: 443,
    auth: keyId + ':' + keySecret
  };
  
  before(function() {
    bitx = new BitX(keyId, keySecret);
    request = sinon.stub(https, 'request'),
    stub = request.withArgs(sinon.match(expectedOptions));
  });
  
  after(function() {
    request.restore();
  });
  
  afterEach(function() {
    stub.reset();
  });
  
  describe('_request', function() {
    
    it('should return the expected result', function(done) {
      var expectedResult = {success: true};
      stub.returns(new FakeRequest(expectedResult));
      bitx._request('GET', path, function(err, result) {
        expect(err).to.be.null;
        expect(result).to.eql(expectedResult);
        done();
      });
    });
    
    it('should return an error if request emits an error', function(done) {
      stub.returns(new FakeRequest(null, {fail: true}));
      bitx._request('GET', path, function(err, result) {
        expect(err).to.be.instanceOf(Error);
        done();
      });
    });
    
    it('should return an error if the response is not valid', function(done) {
      stub.returns(new FakeRequest('invalid'));
      bitx._request('GET', path, function(err, result) {
        expect(err).to.be.instanceOf(Error);
        done();
      });
    });
    
    it('should return an error if the response contains an error', function(done) {
      stub.returns(new FakeRequest({error: true}));
      bitx._request('GET', path, function(err, result) {
        expect(err).to.be.instanceOf(Error);
        done();
      });
    });
    
    it('should return an error if the response is unauthorized', function(done) {
      stub.returns(new FakeRequest(null, {statusCode: 401}));
      bitx._request('GET', path, function(err, result) {
        expect(err).to.be.instanceOf(Error);
        done();
      });
    });
  });
});

describe('External', function() {
  var bitx,
      mock;
  
  var callback = function() {};
  
  before(function() {
    bitx = new BitX();
  });
  
  beforeEach(function() {
    mock = sinon.mock(bitx);
  });
  
  describe('getTicker', function() {
    
    it('should call _request with the correct parameters', function() {
      mock.expects('_request').once().withArgs('GET', 'ticker', callback);
      bitx.getTicker(callback);
      mock.verify();
    });
  });
  
  describe('getOrderBook', function() {
    
    it('should call _request with the correct parameters', function() {
      mock.expects('_request').once().withArgs('GET', 'orderbook', callback);
      bitx.getOrderBook(callback);
      mock.verify();
    });
  });
  
  describe('getTrades', function() {
    
    it('should call _request with the correct parameters', function() {
      mock.expects('_request').once().withArgs('GET', 'trades', callback);
      bitx.getTrades(callback);
      mock.verify();
    });
  });
  
  describe('getOrderList', function() {
    
    it('should call _request with the correct parameters', function() {
      mock.expects('_request').once().withArgs('GET', 'listorders', callback);
      bitx.getOrderList(callback);
      mock.verify();
    });
  });
  
  describe('getLimits', function() {
    
    it('should call _request with the correct parameters', function() {
      mock.expects('_request').once().withArgs('GET', 'getlimits', callback);
      bitx.getLimits(callback);
      mock.verify();
    });
  });
});


test/UnitTests.js
'use strict'

var BitX = require('../lib/BitX')
var FakeRequest = require('./helpers/FakeRequest')
var https = require('https')
var sinon = require('sinon')
var tap = require('tap')

tap.test('BitX constructor', {autoend: true}, function (t) {
  t.test('should create a new instance', function (tt) {
    var bitx = new BitX()
    tt.type(bitx, 'BitX')
    tt.end()
  })

  tap.test('should create a new instance without the new keyword', function (tt) {
    var bitx = BitX()
    tt.type(bitx, 'BitX')
    tt.end()
  })

  tap.test('should have default options', function (tt) {
    var bitx = new BitX()
    tt.equal(bitx.hostname, 'api.luno.com')
    tt.equal(bitx.port, 443)
    tt.equal(bitx.pair, 'XBTZAR')
    tt.end()
  })

  tap.test('should accept options', function (tt) {
    var options = {
      hostname: 'localhost',
      port: 8000,
      pair: 'XBTUSD'
    }
    var bitx = new BitX(options)
    tt.equal(bitx.hostname, options.hostname)
    tt.equal(bitx.port, options.port)
    tt.equal(bitx.pair, options.pair)
    tt.end()
  })

  tap.test('should accept auth and options', function (tt) {
    var keyId = 'cnz2yjswbv3jd'
    var keySecret = '0hydMZDb9HRR3Qq-iqALwZtXLkbLR4fWxtDZvkB9h4I'
    var options = {
      hostname: 'localhost',
      port: 8000,
      pair: 'XBTUSD'
    }
    var bitx = new BitX(keyId, keySecret, options)
    tt.equal(bitx.hostname, options.hostname)
    tt.equal(bitx.port, options.port)
    tt.equal(bitx.pair, options.pair)
    tt.equal(bitx.auth, keyId + ':' + keySecret)
    tt.end()
  })
})

tap.test('Internal', {autoend: true}, function (t) {
  var keyId = '12345'
  var keySecret = '0000000000000000'
  var path = '/api/1/test'

  var expectedOptions = {
    headers: {
      'Accept': 'application/json',
      'Accept-Charset': 'utf-8'
    },
    hostname: 'api.luno.com',
    path: path,
    port: 443,
    auth: keyId + ':' + keySecret
  }

  var bitx = new BitX(keyId, keySecret)
  var request = sinon.stub(https, 'request')
  var stub = request.withArgs(sinon.match(expectedOptions))

  t.afterEach(function () {
    stub.reset()
  })

  t.teardown(function () {
    request.restore()
  })

  t.test('_request should return the expected result', function (tt) {
    var expectedResult = {success: true}
    stub.returns(new FakeRequest(expectedResult))
    bitx._request('GET', path, null, function (err, result) {
      tt.error(err)
      tt.same(result, expectedResult)
      tt.end()
    })
  })

  t.test('_request should return an error if request emits an error', function (tt) {
    stub.returns(new FakeRequest(null, {fail: true}))
    bitx._request('GET', path, null, function (err) {
      tt.type(err, 'Error')
      tt.end()
    })
  })

  t.test('_request should return an error if the response is not valid', function (tt) {
    stub.returns(new FakeRequest('invalid', {stringify: false}))
    bitx._request('GET', path, null, function (err) {
      tt.type(err, 'Error')
      tt.end()
    })
  })

  t.test('_request should return an error if the response contains an error', function (tt) {
    stub.returns(new FakeRequest({error: true}))
    bitx._request('GET', path, null, function (err) {
      tt.type(err, 'Error')
      tt.end()
    })
  })

  t.test('_request should return an error if the response is unauthorized', function (tt) {
    stub.returns(new FakeRequest(null, {statusCode: 401}))
    bitx._request('GET', path, null, function (err) {
      tt.type(err, 'Error')
      tt.end()
    })
  })
})

tap.test('External', {autoend: true}, function (t) {
  var bitx = new BitX()
  var callback = function () {}
  var mock

  t.beforeEach(function () {
    mock = sinon.mock(bitx)
  })

  t.afterEach(function () {
    mock.restore()
  })

  t.test('getTicker should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/ticker', {pair: 'XBTZAR'}, callback)
    bitx.getTicker(callback)
    mock.verify()
    tt.end()
  })

  t.test('getTicker should accept a pair option', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/ticker', {pair: 'XBTMYR'}, callback)
    bitx.getTicker({pair: 'XBTMYR'}, callback)
    mock.verify()
    tt.end()
  })

  t.test('getAllTickers should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/tickers', null, callback)
    bitx.getAllTickers(callback)
    mock.verify()
    tt.end()
  })

  t.test('getOrderBook should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/orderbook', {pair: 'XBTZAR'}, callback)
    bitx.getOrderBook(callback)
    mock.verify()
    tt.end()
  })

  t.test('getOrderBook should accept a pair option', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/orderbook', {pair: 'XBTMYR'}, callback)
    bitx.getOrderBook({pair: 'XBTMYR'}, callback)
    mock.verify()
    tt.end()
  })

  t.test('getTrades should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/trades', {pair: 'XBTZAR'}, callback)
    bitx.getTrades(callback)
    mock.verify()
    tt.end()
  })

  t.test('getTrades should accept a pair option', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/trades', {pair: 'XBTMYR'}, callback)
    bitx.getTrades({pair: 'XBTMYR'}, callback)
    mock.verify()
    tt.end()
  })

  t.test('getOrderList should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/listorders', {pair: 'XBTZAR'}, callback)
    bitx.getOrderList(callback)
    mock.verify()
    tt.end()
  })

  t.test('getOrderList should accept a pair option', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/listorders', {pair: 'XBTMYR'}, callback)
    bitx.getOrderList({pair: 'XBTMYR'}, callback)
    mock.verify()
    tt.end()
  })

  t.test('getLimits should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/BTCZAR/getlimits', null, callback)
    bitx.getLimits(callback)
    mock.verify()
    tt.end()
  })

  t.test('postBuyOrder should call _request with the correct parameters', function (tt) {
    var parameters = {
      type: 'BID',
      volume: 9999.99,
      price: 0.001,
      pair: 'XBTZAR'
    }
    mock.expects('_request').once().withArgs('POST', '/api/1/postorder', parameters, callback)
    bitx.postBuyOrder(parameters.volume, parameters.price, callback)
    mock.verify()
    tt.end()
  })

  t.test('postMarketBuyOrder should call _request with the correct parameters', function (tt) {
    var parameters = {
      type: 'BUY',
      counter_volume: 9999.99,
      pair: 'XBTZAR'
    }
    mock.expects('_request').once().withArgs('POST', '/api/1/marketorder', parameters, callback)
    bitx.postMarketBuyOrder(parameters.counter_volume, callback)
    mock.verify()
    tt.end()
  })

  t.test('postSellOrder should call _request with the correct parameters', function (tt) {
    var parameters = {
      type: 'ASK',
      volume: 0.001,
      price: 9999.99,
      pair: 'XBTZAR'
    }
    mock.expects('_request').once().withArgs('POST', '/api/1/postorder', parameters, callback)
    bitx.postSellOrder(parameters.volume, parameters.price, callback)
    mock.verify()
    tt.end()
  })

  t.test('postMarketSellOrder should call _request with the correct parameters', function (tt) {
    var parameters = {
      type: 'SELL',
      base_volume: 9999.99,
      pair: 'XBTZAR'
    }
    mock.expects('_request').once().withArgs('POST', '/api/1/marketorder', parameters, callback)
    bitx.postMarketSellOrder(parameters.base_volume, callback)
    mock.verify()
    tt.end()
  })

  t.test('stopOrder should call _request with the correct parameters', function (tt) {
    var body = {order_id: 'BXMC2CJ7HNB88U4'}
    mock.expects('_request').once().withArgs('POST', '/api/1/stoporder', body, callback)
    bitx.stopOrder(body.order_id, callback)
    mock.verify()
    tt.end()
  })

  t.test('getBalance should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/balance', null, callback)
    bitx.getBalance(callback)
    mock.verify()
    tt.end()
  })

  t.test('getBalance should accept an asset argument and call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/balance', {asset: 'ZAR'}, callback)
    bitx.getBalance('ZAR', callback)
    mock.verify()
    tt.end()
  })

  t.test('should call _request with the correct parameters', function (tt) {
    mock.expects('_request').once().withArgs('GET', '/api/1/funding_address', {asset: 'XBT'}, callback)
    bitx.getFundingAddress('XBT', callback)
    mock.verify()
    tt.end()
  })
})


test/Regression.js
'use strict'

var BitX = require('..')
var https = require('https')
var fs = require('fs')
var path = require('path')
var tap = require('tap')

tap.test('GET after POST', (t) => {
  var options = {
    key: fs.readFileSync(path.join(__dirname, 'ssl', 'server', 'test.key')),
    cert: fs.readFileSync(path.join(__dirname, 'ssl', 'server', 'test.crt'))
  }
  var server = https.createServer(options).listen(process.env.PORT || 0, 'localhost', function () {
    var address = server.address()
    var bitx = new BitX('keyId', 'keySecret', {
      hostname: 'localhost',
      port: address.port,
      ca: fs.readFileSync(path.join(__dirname, 'ssl', 'ca', 'root.pem'))
    })

    t.afterEach(function () {
      server.removeAllListeners('request')
    })

    t.teardown(function (done) {
      server.close(done)
    })

    t.test('should not send content type or content length headers', function (t) {
      server.once('request', function (req, res) {
        req.resume()
        req.on('end', function () {
          res.end(JSON.stringify({}))
        })
      })

      bitx.postBuyOrder(9999.99, 0.0001, function (err, order) {
        t.error(err)

        server.once('request', function (req, res) {
          t.notOk(req.headers['content-type'])
          t.notOk(req.headers['content-length'])
          res.end(JSON.stringify({}))
        })

        bitx.getTicker(function (err, ticker) {
          t.error(err)
          t.end()
        })
      })
    })

    t.end()
  })
})


lib/BitX.js
'use strict'

var https = require('https')
var path = require('path')
var querystring = require('querystring')
var config = require(path.join(__dirname, '..', 'package'))

var defaultHeaders = {
  'Accept': 'application/json',
  'Accept-Charset': 'utf-8',
  'User-Agent': 'node-bitx v' + config.version
}

function BitX (keyId, keySecret, options) {
  if (!(this instanceof BitX)) {
    return new BitX(keyId, keySecret, options)
  }
  if (typeof keyId === 'string') {
    this.auth = keyId + ':' + keySecret
  } else {
    options = keyId
  }
  options = options || {}
  this.hostname = options.hostname || 'api.luno.com'
  this.port = options.port || 443
  this.ca = options.ca
  this.pair = options.pair || 'XBTZAR'
  this._requestMts = []
  this._requestsRefresh = this._requestsRefresh.bind(this)
  this._timeoutId = null
}

Object.defineProperty(BitX.prototype, 'apiCallRate', {
  get: function () {
    return this._requestsRefresh()
  }
})

BitX.prototype._requestsRefresh = function () {
  var maxMts = Date.now() - 60000
  while (this._requestMts.length > 0) {
    if (this._requestMts[0] < maxMts) {
      this._requestMts.shift()
    } else {
      break
    }
  }

  if (this._requestMts.length > 0) {
    this._timeoutId.refresh()
    this._timeoutId.unref()
  } else {
    clearTimeout(this._timeoutId)
    this._timeoutId = null
  }

  return this._requestMts.length
}

BitX.prototype._request = function (method, resourcePath, data, callback) {
  var headers = Object.assign({}, defaultHeaders)
  data = querystring.stringify(data)
  if (method === 'GET') {
    if (data) {
      resourcePath += '?' + data
    }
  } else if (method === 'POST') {
    headers['Content-Type'] = 'application/x-www-form-urlencoded'
    headers['Content-Length'] = Buffer.byteLength(data)
  }
  var options = {
    headers: headers,
    hostname: this.hostname,
    path: resourcePath,
    port: this.port,
    auth: this.auth,
    method: method
  }
  if (this.ca) {
    options.ca = this.ca
    options.agent = new https.Agent(options)
  }

  var mts = Date.now()
  var requestMts = this._requestMts
  requestMts.push(mts)
  if (this._timeoutId === null) {
    this._timeoutId = setTimeout(this._requestsRefresh, 60000)
    this._timeoutId.unref()
  }

  var promise = new Promise(function (resolve, reject) {
    var req = https.request(options)

    req.on('response', function (res) {
      var response = ''

      res.setEncoding('utf8')

      res.on('data', function (data) {
        response += data
      })

      res.on('end', function () {
        if (res.statusCode !== 200) {
          var lunoApiError
          try {
            response = JSON.parse(response)
            lunoApiError = new Error('luno API error ' + response.error_code + ': ' + response.error)
            lunoApiError.error_code = response.error_code
            if (response.error_code.includes('429') || response.error_code.includes('ErrTooManyRequests')) {
              var i = requestMts.length - 1
              while (i >= 0) {
                if (requestMts[i] === mts) {
                  requestMts.splice(i, 1)
                  break
                }
                i -= 1
              }
            }
          } catch (err) {
            lunoApiError = new Error('luno API error ' + res.statusCode + ': ' + response)
          }
          if (lunoApiError) {
            if (typeof callback === 'function') {
              return callback(lunoApiError)
            } else {
              return reject(lunoApiError)
            }
          }
        }
        try {
          response = JSON.parse(response)
        } catch (err) {
          if (typeof callback === 'function') {
            return callback(err)
          } else {
            return reject(err)
          }
        }
        if (response.error) {
          var err = new Error(response.error)
          if (typeof callback === 'function') {
            return callback(err)
          } else {
            return reject(err)
          }
        }
        if (typeof callback === 'function') {
          return callback(null, response)
        } else {
          return resolve(response)
        }
      })
    })

    req.on('error', function (err) {
      if (typeof callback === 'function') {
        return callback(err)
      } else {
        return reject(err)
      }
    })

    if (method === 'POST' && data) {
      req.write(data)
    }

    req.end()
  })

  if (typeof callback !== 'function') {
    return promise
  }
}

BitX.prototype.getTicker = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  var defaults = {
    pair: this.pair
  }
  return this._request('GET', '/api/1/ticker', Object.assign(defaults, options), callback)
}

BitX.prototype.getAllTickers = function (callback) {
  return this._request('GET', '/api/1/tickers', null, callback)
}

BitX.prototype.getOrderBook = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  var defaults = {
    pair: this.pair
  }
  return this._request('GET', '/api/1/orderbook', Object.assign(defaults, options), callback)
}

BitX.prototype.getTrades = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  var defaults = {
    pair: this.pair
  }
  return this._request('GET', '/api/1/trades', Object.assign(defaults, options), callback)
}

BitX.prototype.getOrderList = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  var defaults = {
    pair: this.pair
  }
  return this._request('GET', '/api/1/listorders', Object.assign(defaults, options), callback)
}

BitX.prototype.getOrderListV2 = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  var defaults = {
    pair: this.pair
  }
  return this._request('GET', '/api/exchange/2/listorders', Object.assign(defaults, options), callback)
}

BitX.prototype.getTradeList = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  var defaults = {
    pair: this.pair
  }
  return this._request('GET', '/api/1/listtrades', Object.assign(defaults, options), callback)
}

BitX.prototype.getLimits = function (callback) {
  console.log('node-bitx warning: BitX.getLimits is deprecated. Please use BitX.getBalance instead.')
  return this._request('GET', '/api/1/BTCZAR/getlimits', null, callback)
}

BitX.prototype.getFeeInfo = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  var defaults = {
    pair: this.pair
  }
  return this._request('GET', '/api/1/fee_info', Object.assign(defaults, options), callback)
}

BitX.prototype.stopOrder = function (orderId, callback) {
  var body = {
    order_id: orderId
  }
  return this._request('POST', '/api/1/stoporder', body, callback)
}

BitX.prototype.postBuyOrder = function (volume, price, options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  var body = {
    type: 'BID',
    volume: volume,
    price: price,
    pair: this.pair
  }
  return this._request('POST', '/api/1/postorder', Object.assign(body, options), callback)
}

BitX.prototype.postMarketBuyOrder = function (volume, options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  var body = {
    type: 'BUY',
    counter_volume: volume,
    pair: this.pair
  }
  return this._request('POST', '/api/1/marketorder', Object.assign(body, options), callback)
}

BitX.prototype.postSellOrder = function (volume, price, options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  var body = {
    type: 'ASK',
    volume: volume,
    price: price,
    pair: this.pair
  }
  return this._request('POST', '/api/1/postorder', Object.assign(body, options), callback)
}

BitX.prototype.postMarketSellOrder = function (volume, options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  var body = {
    type: 'SELL',
    base_volume: volume,
    pair: this.pair
  }
  return this._request('POST', '/api/1/marketorder', Object.assign(body, options), callback)
}

BitX.prototype.getOrder = function (id, callback) {
  return this._request('GET', '/api/1/orders/' + id, null, callback)
}

BitX.prototype.getOrderV2 = function (id, callback) {
  return this._request('GET', '/api/exchange/2/orders/' + id, null, callback)
}

BitX.prototype.getOrderV3 = function (options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  return this._request('GET', '/api/exchange/3/order', options, callback)
}

BitX.prototype.getBalance = function (asset, callback) {
  if (typeof asset === 'function') {
    callback = asset
    asset = null
  }
  return this._request('GET', '/api/1/balance', asset ? {asset: asset} : null, callback)
}

BitX.prototype.getFundingAddress = function (asset, options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  var defaults = {
    asset: asset
  }
  return this._request('GET', '/api/1/funding_address', Object.assign(defaults, options), callback)
}

BitX.prototype.createFundingAddress = function (asset, callback) {
  return this._request('POST', '/api/1/funding_address', {asset: asset}, callback)
}

BitX.prototype.getTransactions = function (asset, options, callback) {
  if (typeof options === 'function') {
    callback = options
    options = {}
  }
  var defaults = {
    asset: asset,
    offset: 0,
    limit: 10
  }
  return this._request('GET', '/api/1/transactions', Object.assign(defaults, options), callback)
}

BitX.prototype.getWithdrawals = function (callback) {
  return this._request('GET', '/api/1/withdrawals/', null, callback)
}

BitX.prototype.getWithdrawal = function (id, callback) {
  return this._request('GET', '/api/1/withdrawals/' + id, null, callback)
}

BitX.prototype.requestWithdrawal = function (type, amount, callback) {
  var options = {
    type: type,
    amount: amount
  }
  return this._request('POST', '/api/1/withdrawals/', options, callback)
}

BitX.prototype.cancelWithdrawal = function (id, callback) {
  return this._request('DELETE', '/api/1/withdrawals/' + id, null, callback)
}

module.exports = BitX
